### PR TITLE
Feature/review

### PR DESCRIPTION
**Pull request by Ponchimeow**
layout beta, will change a little design when all data concatenation.

### CODE REVIEW

File: src/components/common/CommonModal.tsx [hunk 1]
```diff
@@ -13,17 +13,19 @@ import React from 'react'
 const CommonModal: React.FC<
   ModalProps & {
     title: string
+    renderHeaderIcon?: () => React.ReactElement
     renderTrigger: () => React.ReactElement
     renderFooter?: () => React.ReactElement
```
Comment (futianshen) [inline src/components/common/CommonModal.tsx:18]:
```js
title: string
renderTrigger: () => React.ReactElement
renderHeaderIcon?: () => React.ReactElement
renderFooter?: () => React.ReactElement
```

File: src/components/common/CommonModal.tsx [hunk 2]
```diff
@@ -13,17 +13,19 @@ import React from 'react'
...
-> = ({ title, renderTrigger, renderFooter, children, ...ModalProps }) => {
+> = ({ title, renderHeaderIcon, renderTrigger, renderFooter, children, ...ModalProps }) => {
   return (
     <>
       {renderTrigger()}
 
       <Modal {...ModalProps}>
         <ModalOverlay />
         <ModalContent>
+          {renderHeaderIcon && renderHeaderIcon()}
```
Comment (futianshen) [inline src/components/common/CommonModal.tsx:28]:
```jsx
{renderHeaderIcon?.()}
```

File: src/components/review/ReviewCollectionBlock.tsx [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+  const { formatMessage } = useIntl()
+  const { currentMemberId, currentUserRole } = useAuth()
+  const { settings } = useApp()
+
+  const { reviews, reviewCount } = useReviewCollection(path)
+
+  const AverageScore = (reviews: ReviewsProps[]) => {
+    let totalScore = 0
+    reviews.map(v => (totalScore += v.score))
+    return reviewCount === 0 ? 0 : (totalScore / reviewCount).toFixed(1)
```
Comment (futianshen) [inline src/components/review/ReviewCollectionBlock.tsx:73]: reviewCount && (totalScore / reviewCount).toFixed(1)

File: src/components/review/ReviewReplyItem.tsx [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+  const toast = useToast()
+  const [replyEditing, setReplyEditing] = useState(false)
+  const [isSubmitting, setIsSubmitting] = useState(false)
+
+  const [updateReviewReply] = useMutation<types.UPDATE_REVIEW_REPLY, types.UPDATE_REVIEW_REPLYVariables>(
+    UPDATE_REVIEW_REPLY,
+  )
+  const [deleteReviewReply] = useMutation<types.DELETE_REVIEW_REPLY, types.DELETE_REVIEW_REPLYVariables>(
+    DELETE_REVIEW_REPLY,
+  )
```
Comment (futianshen) [inline src/components/review/ReviewReplyItem.tsx:96]: can use a custom hook to wrap it up
```
const { updateReviewReply, deleteReviewReply } = useMutateReviewReply(reviewReplyId, currentMemberId, appId)
```

**Comment by Elantris** [comment]
[(8) 【評價】評價前台切版 [6]](https://trello.com/c/D1G3u4rl/1566-8-%E3%80%90%E8%A9%95%E5%83%B9%E3%80%91%E8%A9%95%E5%83%B9%E5%89%8D%E5%8F%B0%E5%88%87%E7%89%88-6)

File: src/components/review/ReviewAdminItem.tsx [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+        appId,
+        targetId,
+        limit: 5,
+      },
+    },
+  )
+
+  const reviews: ReviewProps[] =
+    data?.review.map(v => ({
+      reviewId: v.id,
```
Comment (futianshen) [inline src/components/review/ReviewAdminItem.tsx:121]: just use id

File: src/components/review/ReviewAdminItem.tsx [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+  const { loadingReviews, reviews, labelRole, onRefetch, loadMoreReviews } = useReviewCollection(path, appId, targetId)
+
+  return (
+    <>
+      {reviews.map((v: ReviewProps) => {
+        return (
+          <div key={uuid()} className="review-item">
+            <ReviewItem
+              isAdmin
+              key={uuid()}
```
Comment (futianshen) [inline src/components/review/ReviewAdminItem.tsx:28]: why not use reviewId ?

File: src/components/review/ReviewMemberItem.tsx [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+      },
+    })
+
+  return {
+    loadingReviews: loading,
+    error,
+    memberReviews,
+    memberPrivateContent,
+    labelRole,
+    onRefetch: refetch,
```
Comment (futianshen) [inline src/components/review/ReviewMemberItem.tsx:215]: use `refetch` or `refetchReviews`

File: src/components/review/ReviewAdminItem.tsx [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+              onRefetch={onRefetch}
+            />
+            <Divider
+              className="review-divider"
+              css={{ margin: '24px 0', height: '1px', background: '#ececec', borderStyle: 'none', opacity: 1 }}
+            />
+          </div>
+        )
+      })}
+      {!loadingReviews && loadMoreReviews && (
```
Comment (futianshen) [inline src/components/review/ReviewAdminItem.tsx:48]: can add loading state before loading complete
```js
if (loadingReviews) {
  return
}

return  <>
      {reviews.map((v: ReviewProps) => { ...

```

File: src/components/review/ReviewAdminItem.tsx [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+}> = ({ targetId, path, appId }) => {
+  const { formatMessage } = useIntl()
+  const [loading, setLoading] = useState(false)
+  const { loadingReviews, reviews, labelRole, onRefetch, loadMoreReviews } = useReviewCollection(path, appId, targetId)
+
+  return (
+    <>
+      {reviews.map((v: ReviewProps) => {
+        return (
+          <div key={uuid()} className="review-item">
```
Comment (futianshen) [inline src/components/review/ReviewAdminItem.tsx:25]:
```js
{reviews.map(v => (
   <div key={uuid()} className="review-item">
     ...
   </div>
)}
```

File: src/components/review/ReviewAdminItem.tsx [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+        })
+      },
+    })
+
+  return {
+    loadingReviews: loading,
+    error,
+    reviews,
+    labelRole,
+    onRefetch: refetch,
```
Comment (futianshen) [inline src/components/review/ReviewAdminItem.tsx:171]: Keep the naming consistent
```js
{
   loadingReviews: loading,
   reviews,
   errorReviews: error,
   refetchReviews: refetch
}
```

File: src/helpers/translation.ts [hunk 11]
```diff
@@ -1275,6 +1277,44 @@ export const projectMessages = {
...
+    edited: { id: 'review.status.edited', defaultMessage: '已編輯' },
+  }),
+  text: defineMessages({
+    reviewAmount: { id: 'review.text.amount', defaultMessage: '{amount} 則評價' },
+    notEnoughReviews: { id: 'review.text.notEnoughReviews', defaultMessage: '評價不足三人無法顯示' },
+    reply: { id: 'review.text.reply', defaultMessage: '回覆...' },
+    reviewModalDescription: {
+      id: 'review.text.reviewModalDescription',
+      defaultMessage: '歡迎給予鼓勵、分享心得或是提出建議！為維護評價公正性，累計三則以上評價才會公開呦！',
+    },
```
Comment (Elantris) [inline src/helpers/translation.ts:1294]: the count in messages should be variables

File: src/types/review.ts [hunk 12]
```diff
@@ -0,0 +1,40 @@
+export type reviewRoleName = 'app-owner' | 'content-creator'
```
Comment (Elantris) [inline src/types/review.ts:1]: remove unused type definition

File: src/types/review.ts [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+export type MemberReviewProps = {
+  id: string
+  memberId: string
+  score: number
+  title: string
+  content: string | null
+  privateContent: string | null
+} | null
```
Comment (Elantris) [inline src/types/review.ts:40]: type definition of objects should not union with `null`

File: src/components/review/ReviewCollectionBlock.tsx [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+  const { settings, id: appId } = useApp()
+  const { averageScore, reviewCount, productRoles, reviewCountAndProductRolesRefetch } = useReviewCountAndProductRoles(
+    targetId,
+    path,
+    appId,
+  )
+  const { enrolledMembers } = useEnrolledMembers(targetId, appId)
+  const { memberReview } = useMemberReview(currentMemberId, path, appId)
+
+  const reviewMemberItemRef = useRef() as React.RefObject<ReviewMemberItemRef>
```
Comment (Elantris) [inline src/components/review/ReviewCollectionBlock.tsx:75]: NITS: interesting flow of ref 🤔

**Review by Elantris — CHANGES_REQUESTED** [review]

File: src/helpers/index.ts [hunk 15]
```diff
@@ -311,3 +311,8 @@ export const validationRegExp: { [fieldId: string]: RegExp } = {
   citizenCode: /^[a-zA-Z]{2}[0-9]{14}$/,
   uniformNumber: /^[0-9]{8}$/,
 }
+
+export const numberToText = (value: number) => {
+  const twNum = '零一二三四五六七八九'
+  return twNum[value]
+}
```
Comment (Elantris) [inline src/helpers/index.ts:318]: don't use static string of specific language in translation

**Pull request merged**